### ISSUE TITLE
OpenMLRun.get_metric_fn does not use predicted probabilities

### ISSUE THREAD

**Issue by PGijsbers** [openml/openml-python#754]
When calling `OpenMLRun.get_metric_fn` with a score function predicted labels are always used, even if the metric should prefer predicted probabilities. See below for code demonstrating the effect for `roc_auc_score`. 

``` python
>>> import numpy as np
>>> import openml
>>> from sklearn.metrics import roc_auc_score
>>> from sklearn.ensemble import RandomForestClassifier
>>> rf = RandomForestClassifier(n_estimators=100)
>>> task = openml.tasks.get_task(146228) # solarflare
>>> run = openml.runs.run_model_on_task(rf, task)
>>> data = np.asarray(run.data_content)
>>> for fold in range(10):
...     fold_idx = data[:, 1] == str(fold)
...     truth = data[fold_idx, -1].astype(int)
...     proba = data[fold_idx, -3].astype(float)
...     pred = data[fold_idx, -2].astype(float)
...     print('proba:', roc_auc_score(truth, proba))
...     print('pred:', roc_auc_score(truth, pred))
...
proba: 0.8295880149812734
pred: 0.6385767790262172
proba: 0.5777153558052435
pred: 0.49968789013732834
proba: 0.7159800249687891
pred: 0.6607365792759051
proba: 0.6885143570536829
pred: 0.5995630461922596
proba: 0.6722488038277512
pred: 0.646531100478469
proba: 0.5487440191387559
pred: 0.4808612440191387
proba: 0.7307449494949495
pred: 0.6325757575757576
proba: 0.5912247474747474
pred: 0.5157828282828283
proba: 0.7585227272727273
pred: 0.6212121212121211
proba: 0.5997474747474748
pred: 0.5599747474747475
>>> run.get_metric_fn(roc_auc_score)
array([0.63857678, 0.49968789, 0.66073658, 0.59956305, 0.6465311 ,
       0.48086124, 0.63257576, 0.51578283, 0.62121212, 0.55997475])
>>> data  # for reference
array([['0', '0', '0', ..., '0.0', '0', '0'],
       ['0', '0', '0', ..., '0.035', '0', '0'],
       ['0', '0', '0', ..., '0.07860197211379191', '0', '0'],
       ...,
       ['0', '9', '0', ..., '0.05442641202816864', '0', '1'],
       ['0', '9', '0', ..., '0.0', '0', '1'],
       ['0', '9', '0', ..., '0.2425', '0', '1']], dtype='<U32')
```

**Comment by mfeurer** [comment]
Is this related to #386?

**Comment by PGijsbers** [comment]
I think they are two different issues.

As I understood #386, when calculating the score on a learning curve task, *all* predictions, including those made from models trained on subsamples of the data, are used to calculate the final score. Fixing this requires taking into account the `sample` column (and calculating the score for a specified sample number).

Here, for any task scores are computed with predicted labels (also tasks that do not include subsamples), even when the metric should prefer to use the predicted class probabilities instead. Fixing this would require us to know which metric prefers class probabilities over predicted labels, and then using the probabilities as opposed to the labels.

**Comment by amueller** [comment]
This can't really be done with the function interface. This could be done with the ``scorer`` interface. These are callables that know how to get the right predictions from the models. But the interface is ```auc_scorer(estimator, X, y)`` and I assume we don't have the estimator at this point and only have the predicted probabilities.
We could get the data about whether to use probabilities out of the scorer internals but that's not future proof as it's not a public API.

The docs say ``Calculates metric scores based on predicted values`` so I think this is expected behavior for ``get_metric_fn``. But it's clearly not ideal.

If this is for sklearn specific scorers, you should have users specify the scorers by string and use ``get_scorer``. You can then fake having the estimator and use the scorer API.

I'd say I'm happy to help but I have no idea what ``get_metric_fn`` does.

**Comment by PGijsbers** [comment]
@mfeurer The docstring says this function assumes the run has been executed locally.. can we then assume we also still have the estimator? Though then that would make this function scikit-learn specific I suppose.

 What do you think we ought to do here? As is, I'm not a big fan of keeping this if metrics such as `roc_auc_score` are computed with predicted labels over probabilities.

**Comment by mfeurer** [comment]
> The docstring says this function assumes the run has been executed locally.. can we then assume we also still have the estimator?

I'm not sure about this. I would imagine this to work with downloaded runs, too. Could you give that a try?

However, I'm not 100% sure what this functions intended use is. @janvanrijn I think you added that back then, could you please explain what we need this for?

Anyway, I like @amueller's suggestion: "If this is for sklearn specific scorers, you should have users specify the scorers by string and use get_scorer. You can then fake having the estimator and use the scorer API."
